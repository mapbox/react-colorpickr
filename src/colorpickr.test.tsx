import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ColorPickr from './colorpickr';
import CopyButton from '@mapbox/mr-ui/copy-button';
import '@testing-library/jest-dom/extend-expect';

describe('Colorpickr', () => {
  const user = userEvent.setup();

  test('renders', () => {
    const props = {
      onChange: jest.fn()
    };

    const { baseElement } = render(<ColorPickr {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  describe('basics', () => {
    const props = {
      onChange: jest.fn()
    };

    beforeEach(() => {
      render(<ColorPickr {...props} />);
    });

    test('color input returns value onChange', () => {
      const mockEvent = {
        target: {
          value: 'eeef'
        }
      };

      const input = screen.getByTestId('color-input');
      fireEvent.change(input, mockEvent);
      expect(props.onChange).toHaveBeenCalledTimes(1);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 0,
        l: 93,
        r: 238,
        g: 238,
        b: 238,
        a: 1,
        hex: 'eeeeee',
        colorSpace: 'hex',
        mode: 'disc'
      });
    });

    test('invalid color input does not fire onChange', () => {
      const mockEvent = {
        target: {
          value: 'eeeff'
        }
      };

      const input = screen.getByTestId('color-input');
      fireEvent.change(input, mockEvent);
      expect(props.onChange).toHaveBeenCalledTimes(0);
    });

    test('leading hash is permitted', () => {
      const mockEvent = {
        target: {
          value: '#333'
        }
      };

      const input = screen.getByTestId('color-input');
      fireEvent.change(input, mockEvent);
      expect(props.onChange).toHaveBeenCalledTimes(1);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 0,
        l: 20,
        r: 51,
        g: 51,
        b: 51,
        a: 1,
        hex: '333333',
        colorSpace: 'hex',
        mode: 'disc'
      });
    });

    test('default mode is active', () => {
      const element = screen.getByTestId('mode-disc');
      expect(element.classList.contains('is-active')).toEqual(true);
    });

    test('default reset action is present', () => {
      const element = screen.getByTestId('color-reset');
      expect(element).toBeTruthy();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });
  });

  describe('overrides', () => {
    interface MockedInstance {
      overrideValue: (
        v: string,
        updateInit?: boolean,
        emitOnChange?: boolean
      ) => void;
    }

    let mockedInstance: MockedInstance;

    const props = {
      onChange: jest.fn(),
      // @ts-expect-error accuracy not required for test
      mounted: (instance) => (mockedInstance = instance)
    };

    beforeEach(() => {
      render(<ColorPickr {...props} />);
    });

    test('mocked instance is called', () => {
      mockedInstance.overrideValue('red');
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 100,
        l: 50,
        r: 255,
        g: 0,
        b: 0,
        a: 1,
        hex: 'ff0000',
        colorSpace: 'hex',
        mode: 'disc'
      });

      const element = screen.getByTestId('color-reset');
      expect(element.style.backgroundColor).toEqual('rgb(0, 0, 0)');
    });

    test('overrideValue with true as second argument sets overrides initialValue', () => {
      mockedInstance.overrideValue('red', true);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 100,
        l: 50,
        r: 255,
        g: 0,
        b: 0,
        a: 1,
        hex: 'ff0000',
        colorSpace: 'hex',
        mode: 'disc'
      });

      const element = screen.getByTestId('color-reset');
      expect(element.style.backgroundColor).toEqual('rgb(255, 0, 0)');
    });

    test('emitOnChange: false does not invoke onChange handler', () => {
      mockedInstance.overrideValue('red', true, false);
      expect(props.onChange).not.toHaveBeenCalled();
      const element = screen.getByTestId('color-reset');
      expect(element.style.backgroundColor).toEqual('rgb(255, 0, 0)');
    });

    afterEach(() => {
      jest.clearAllMocks();
    });
  });

  describe('readOnly', () => {
    const props = {
      onChange: jest.fn(),
      readOnly: true
    };

    test('renders', async () => {
      const { baseElement } = render(<ColorPickr {...props} />);
      expect(baseElement).toMatchSnapshot();
      expect(screen.getByTestId('color-input')).toHaveAttribute('readonly');
      expect(screen.getByTestId('color-reset')).toHaveAttribute('disabled');
      expect(screen.queryAllByTestId('color-copy').length).toEqual(0);
      expect(screen.getByTestId('hue-slider')).toHaveAttribute('data-disabled');
      expect(screen.getByTestId('alpha-slider')).toHaveAttribute(
        'data-disabled'
      );

      const input = screen.getByTestId('mode-values');
      await user.click(input);

      expect(screen.getByTestId('h-input')).toHaveAttribute('readonly');
      expect(screen.getByTestId('s-input')).toHaveAttribute('readonly');
      expect(screen.getByTestId('l-input')).toHaveAttribute('readonly');
      expect(screen.getByTestId('r-input')).toHaveAttribute('readonly');
      expect(screen.getByTestId('g-input')).toHaveAttribute('readonly');
      expect(screen.getByTestId('b-input')).toHaveAttribute('readonly');
      expect(screen.getByTestId('a-input')).toHaveAttribute('readonly');
    });
  });

  describe('copy', () => {
    const props = {
      onChange: jest.fn()
    };

    test('Displays copy tooltip', async () => {
      jest.spyOn(CopyButton, 'isCopySupported').mockReturnValue(true);
      render(<ColorPickr {...props} />);
      await user.click(screen.getByTestId('color-copy'));
      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });
    });
  });

  describe('reset:false', () => {
    const props = {
      onChange: jest.fn(),
      reset: false
    };

    test('renders', () => {
      const { queryByTestId } = render(<ColorPickr {...props} />);
      expect(queryByTestId('color-reset')).toBeFalsy();
    });
  });

  describe('reset', () => {
    const props = {
      onChange: jest.fn(),
      alpha: false
    };

    test('renders', () => {
      const { queryByLabelText, queryByTestId } = render(
        <ColorPickr {...props} />
      );
      expect(queryByLabelText('Î±')).toBeFalsy();
      expect(queryByTestId('alpha-slider')).toBeFalsy();
    });
  });

  describe('color input', () => {
    const props = {
      onChange: jest.fn()
    };

    test('hex', () => {
      render(<ColorPickr {...props} />);
      const mockEvent = {
        target: {
          value: '#333'
        }
      };

      fireEvent.change(screen.getByTestId('color-input'), mockEvent);
      expect(props.onChange).toHaveBeenCalledTimes(1);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 0,
        l: 20,
        r: 51,
        g: 51,
        b: 51,
        a: 1,
        hex: '333333',
        colorSpace: 'hex',
        mode: 'disc'
      });
    });

    test('hex: missing hash', () => {
      render(<ColorPickr {...props} />);
      const mockEvent = {
        target: {
          value: '4163fb'
        }
      };

      fireEvent.change(screen.getByTestId('color-input'), mockEvent);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 229,
        s: 96,
        l: 62,
        r: 65,
        g: 99,
        b: 251,
        a: 1,
        hex: '4163fb',
        colorSpace: 'hex',
        mode: 'disc'
      });
    });

    test('rgb', () => {
      render(<ColorPickr {...props} />);
      const mockEvent = {
        target: {
          value: 'rgb(255, 0, 0)'
        }
      };

      fireEvent.change(screen.getByTestId('color-input'), mockEvent);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 100,
        l: 50,
        r: 255,
        g: 0,
        b: 0,
        a: 1,
        hex: 'ff0000',
        colorSpace: 'hex',
        mode: 'disc'
      });
    });

    test('rgba', () => {
      render(<ColorPickr {...props} />);
      const mockEvent = {
        target: {
          value: 'rgb(255, 0, 0, 0.25)'
        }
      };

      fireEvent.change(screen.getByTestId('color-input'), mockEvent);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 0,
        s: 100,
        l: 50,
        r: 255,
        g: 0,
        b: 0,
        a: 0.25,
        hex: 'ff0000',
        colorSpace: 'hex',
        mode: 'disc'
      });
    });

    test('hsl', () => {
      render(<ColorPickr {...props} />);
      const mockEvent = {
        target: {
          value: 'hsla(261, 100%, 50%)'
        }
      };

      fireEvent.change(screen.getByTestId('color-input'), mockEvent);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 261,
        s: 100,
        l: 50,
        r: 89,
        g: 0,
        b: 255,
        a: 1,
        hex: '5900ff',
        colorSpace: 'hex',
        mode: 'disc'
      });
    });

    test('hsla', () => {
      render(<ColorPickr {...props} />);
      const mockEvent = {
        target: {
          value: 'hsla(261, 100%, 50%,0.61)'
        }
      };

      fireEvent.change(screen.getByTestId('color-input'), mockEvent);
      expect(props.onChange).toHaveBeenCalledWith({
        h: 261,
        s: 100,
        l: 50,
        r: 89,
        g: 0,
        b: 255,
        a: 0.61,
        hex: '5900ff',
        colorSpace: 'hex',
        mode: 'disc'
      });
    });

    test('different colorspace outputs expected value', () => {
      render(
        <ColorPickr
          onChange={jest.fn()}
          colorSpace="hsl"
          initialValue="hsla(100, 50%, 40%, 0.5)"
        />
      );
      const input = screen.getByTestId('color-input') as HTMLInputElement;
      expect(input.value).toEqual('hsla(100, 50%, 40%, 0.5)');
    });

    test('color input value adjusts from invalid value to a valid value onBlur', () => {
      render(
        <ColorPickr
          onChange={jest.fn()}
          initialValue="hsla(100, 50%, 40%, 0.5)"
        />
      );
      const mockEvent = {
        target: {
          value: 'eeef'
        }
      };

      const input = screen.getByTestId('color-input') as HTMLInputElement;
      expect(input.value).toEqual('#559933');
      fireEvent.change(input, mockEvent);
      expect(input.value).toEqual('eeef');
      fireEvent.blur(input, mockEvent);
      expect(input.value).toEqual('#eeeeee');
    });
  });

  describe('modes', () => {
    test('returns values on change', async () => {
      const props = {
        initialValue: 'rgba(0, 255, 255, 0.5)',
        onChange: jest.fn()
      };

      render(<ColorPickr {...props} />);
      const input = screen.getByTestId('mode-values');
      await user.click(input);

      await userEvent.selectOptions(
        screen.getByTestId('colorspace-select'),
        'rgb'
      );

      expect(props.onChange).toHaveBeenCalledWith({
        h: 180,
        s: 100,
        l: 50,
        r: 0,
        g: 255,
        b: 255,
        a: 0.5,
        hex: '00ffff',
        colorSpace: 'rgb',
        mode: 'values'
      });
    });
  });
});
