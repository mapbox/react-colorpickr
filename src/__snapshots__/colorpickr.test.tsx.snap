// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colorpickr readOnly renders 1`] = `
<body>
  <div>
    <div
      class="colorpickr round inline-block w240 txt-xs flex flex--column"
    >
      <div
        class="flex-child-no-shrink border-b border--gray-light"
      >
        <button
          class="link link--gray pb6 px3 mb-neg1 is-active border-b border-gray"
          data-state="closed"
          data-testid="mode-disc"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="events-none icon"
            data-testid="icon-circle"
            focusable="false"
            style="width: 18px; height: 18px;"
          >
            <use
              xlink:href="#icon-circle"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            />
          </svg>
          <span
            style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          >
            circle
          </span>
        </button>
        <button
          class="link link--gray pb6 px3 mb-neg1"
          data-state="closed"
          data-testid="mode-values"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="events-none icon"
            data-testid="icon-boolean"
            focusable="false"
            style="width: 18px; height: 18px;"
          >
            <use
              xlink:href="#icon-boolean"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            />
          </svg>
          <span
            style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          >
            boolean
          </span>
        </button>
      </div>
      <div
        class="flex-child-no-shrink mb6 h240 w-full overflow-hidden"
      >
        <div
          class="relative w-full h-full cursor-pointer events-none"
          data-testid="xy"
        >
          <div
            class="absolute z1 round-full border border--2 border--white cursor-move"
            style="background-color: rgb(0, 0, 0); width: 18px; height: 18px; top: calc(100% - 9px); left: calc(0% - 9px);"
          />
          <div
            class="absolute top right bottom left"
            style="background: rgb(255, 0, 0);"
          />
          <div
            class="absolute top right bottom left gradient-hue"
          />
        </div>
      </div>
      <div
        class="flex-child-no-shrink mb6"
      >
        <span
          aria-disabled="true"
          class="relative flex h12 w-full round-full cursor-pointer"
          data-disabled=""
          data-orientation="horizontal"
          data-testid="hue-slider"
          dir="ltr"
          id="hue"
          style="align-items: center; --radix-slider-thumb-transform: translateX(-50%);"
        >
          <div
            class="absolute w-full h3 bg-tile"
          />
          <span
            class="h3 relative flex-child-grow round-full"
            data-disabled=""
            data-orientation="horizontal"
          />
          <span
            style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(0% + 0px);"
          >
            <span
              aria-orientation="horizontal"
              aria-valuemax="360"
              aria-valuemin="0"
              aria-valuenow="0"
              class="w12 h12 round-full block border border--white shadow-darken10"
              data-disabled=""
              data-orientation="horizontal"
              data-radix-collection-item=""
              role="slider"
              style="cursor: ew-resize; background-color: rgb(255, 0, 0);"
            />
          </span>
        </span>
        <span
          aria-disabled="true"
          class="relative flex h12 w-full round-full cursor-pointer"
          data-disabled=""
          data-orientation="horizontal"
          data-testid="alpha-slider"
          dir="ltr"
          id="alpha"
          style="align-items: center; --radix-slider-thumb-transform: translateX(-50%);"
        >
          <div
            class="absolute w-full h3 bg-tile"
          />
          <span
            class="h3 relative flex-child-grow round-full"
            data-disabled=""
            data-orientation="horizontal"
          />
          <span
            style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(100% + 0px);"
          >
            <span
              aria-orientation="horizontal"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="100"
              class="w12 h12 round-full block border border--white shadow-darken10"
              data-disabled=""
              data-orientation="horizontal"
              data-radix-collection-item=""
              role="slider"
              style="cursor: ew-resize; background-color: rgb(0, 0, 0);"
            />
          </span>
        </span>
      </div>
      <div
        class="flex mb12"
      >
        <div
          class="w-1/2 h24 flex flex--center-cross"
        >
          <div
            class="relative w-full"
          >
            <div
              class="select-container w-7/12"
            >
              <select
                aria-required="true"
                class="select txt-s color-gray"
                data-testid="colorspace-select"
                id="colorspace"
              >
                <option
                  value="hsl"
                >
                  HSL
                </option>
                <option
                  value="rgb"
                >
                  RGB
                </option>
                <option
                  value="hex"
                >
                  HEX
                </option>
              </select>
              <div
                class="select-arrow"
              />
            </div>
            <div
              role="alert"
            />
          </div>
        </div>
        <div
          class="w-1/2"
        >
          <input
            class="w-full input input--s"
            data-testid="color-input"
            readonly=""
            type="text"
            value="#000"
          />
        </div>
      </div>
      <div
        class="flex round overflow-hidden"
      >
        <div
          class="bg-tile w-1/2 h24 border-r border--gray-faint"
        >
          <button
            aria-disabled="true"
            aria-label="Reset color"
            class="w-full h-full txt-bold align-center color-transparent color-white-on-hover transition"
            data-testid="color-reset"
            disabled=""
            style="background-color: rgb(0, 0, 0);"
            type="button"
          />
        </div>
        <div
          class="bg-tile w-1/2 h24"
        >
          <div
            class="w-full h-full"
            style="background-color: rgb(0, 0, 0);"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Colorpickr renders 1`] = `
<body>
  <div>
    <div
      class="colorpickr round inline-block w240 txt-xs flex flex--column"
    >
      <div
        class="flex-child-no-shrink border-b border--gray-light"
      >
        <button
          class="link link--gray pb6 px3 mb-neg1 is-active border-b border-gray"
          data-state="closed"
          data-testid="mode-disc"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="events-none icon"
            data-testid="icon-circle"
            focusable="false"
            style="width: 18px; height: 18px;"
          >
            <use
              xlink:href="#icon-circle"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            />
          </svg>
          <span
            style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          >
            circle
          </span>
        </button>
        <button
          class="link link--gray pb6 px3 mb-neg1"
          data-state="closed"
          data-testid="mode-values"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="events-none icon"
            data-testid="icon-boolean"
            focusable="false"
            style="width: 18px; height: 18px;"
          >
            <use
              xlink:href="#icon-boolean"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            />
          </svg>
          <span
            style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
          >
            boolean
          </span>
        </button>
      </div>
      <div
        class="flex-child-no-shrink mb6 h240 w-full overflow-hidden"
      >
        <div
          class="relative w-full h-full cursor-pointer"
          data-testid="xy"
        >
          <div
            class="absolute z1 round-full border border--2 border--white cursor-move"
            style="background-color: rgb(0, 0, 0); width: 18px; height: 18px; top: calc(100% - 9px); left: calc(0% - 9px);"
          />
          <div
            class="absolute top right bottom left"
            style="background: rgb(255, 0, 0);"
          />
          <div
            class="absolute top right bottom left gradient-hue"
          />
        </div>
      </div>
      <div
        class="flex-child-no-shrink mb6"
      >
        <span
          aria-disabled="false"
          class="relative flex h12 w-full round-full cursor-pointer"
          data-orientation="horizontal"
          data-testid="hue-slider"
          dir="ltr"
          id="hue"
          style="align-items: center; --radix-slider-thumb-transform: translateX(-50%);"
        >
          <div
            class="absolute w-full h3 bg-tile"
          />
          <span
            class="h3 relative flex-child-grow round-full"
            data-orientation="horizontal"
          />
          <span
            style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(0% + 0px);"
          >
            <span
              aria-orientation="horizontal"
              aria-valuemax="360"
              aria-valuemin="0"
              aria-valuenow="0"
              class="w12 h12 round-full block border border--white shadow-darken10"
              data-orientation="horizontal"
              data-radix-collection-item=""
              role="slider"
              style="cursor: ew-resize; background-color: rgb(255, 0, 0);"
              tabindex="0"
            />
          </span>
        </span>
        <span
          aria-disabled="false"
          class="relative flex h12 w-full round-full cursor-pointer"
          data-orientation="horizontal"
          data-testid="alpha-slider"
          dir="ltr"
          id="alpha"
          style="align-items: center; --radix-slider-thumb-transform: translateX(-50%);"
        >
          <div
            class="absolute w-full h3 bg-tile"
          />
          <span
            class="h3 relative flex-child-grow round-full"
            data-orientation="horizontal"
          />
          <span
            style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(100% + 0px);"
          >
            <span
              aria-orientation="horizontal"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="100"
              class="w12 h12 round-full block border border--white shadow-darken10"
              data-orientation="horizontal"
              data-radix-collection-item=""
              role="slider"
              style="cursor: ew-resize; background-color: rgb(0, 0, 0);"
              tabindex="0"
            />
          </span>
        </span>
      </div>
      <div
        class="flex mb12"
      >
        <div
          class="w-1/2 h24 flex flex--center-cross"
        >
          <div
            class="relative w-full"
          >
            <div
              class="select-container w-7/12"
            >
              <select
                aria-required="true"
                class="select txt-s color-gray"
                data-testid="colorspace-select"
                id="colorspace"
              >
                <option
                  value="hsl"
                >
                  HSL
                </option>
                <option
                  value="rgb"
                >
                  RGB
                </option>
                <option
                  value="hex"
                >
                  HEX
                </option>
              </select>
              <div
                class="select-arrow"
              />
            </div>
            <div
              role="alert"
            />
          </div>
        </div>
        <div
          class="w-1/2"
        >
          <input
            class="w-full input input--s bg-white"
            data-testid="color-input"
            type="text"
            value="#000"
          />
        </div>
      </div>
      <div
        class="flex round overflow-hidden"
      >
        <div
          class="bg-tile w-1/2 h24 border-r border--gray-faint"
        >
          <button
            aria-label="Reset color"
            class="w-full h-full txt-bold align-center color-transparent color-white-on-hover transition"
            data-state="closed"
            data-testid="color-reset"
            style="background-color: rgb(0, 0, 0);"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="events-none icon inline-block align-t"
              data-testid="icon-undo"
              focusable="false"
              style="width: 18px;"
            >
              <use
                xlink:href="#icon-undo"
                xmlns:xlink="http://www.w3.org/1999/xlink"
              />
            </svg>
            <span
              style="position: absolute; border: 0px; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; word-wrap: normal;"
            >
              undo
            </span>
          </button>
        </div>
        <div
          class="bg-tile w-1/2 h24"
        >
          <div
            class="w-full h-full"
            style="background-color: rgb(0, 0, 0);"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
